File: src/renderer/vulkan/SamplerResources.cpp
////////////////////////////////////////////////////////////////////////////////
/// Reaper
///
/// Copyright (c) 2015-2022 Thibault Schueller
/// This file is distributed under the MIT License
////////////////////////////////////////////////////////////////////////////////

#include "SamplerResources.h"

#include "renderer/vulkan/Backend.h"
#include "renderer/vulkan/Debug.h"
#include "renderer/vulkan/renderpass/ShadowConstants.h"

#include <common/Log.h>
#include <common/ReaperRoot.h>

#include <cfloat>
#include <core/Assert.h>

namespace Reaper
{
namespace
{
    VkSampler create_sampler(VulkanBackend& backend, const char* debug_name, const VkSamplerCreateInfo& create_info)
    {
        VkSampler sampler;
        Assert(vkCreateSampler(backend.device, &create_info, nullptr, &sampler) == VK_SUCCESS);

        VulkanSetDebugName(backend.device, sampler, debug_name);

        return sampler;
    }
} // namespace

SamplerResources create_sampler_resources(ReaperRoot& /*root*/, VulkanBackend& backend)
{
    SamplerResources resources;

    resources.linear_clamp = create_sampler(backend, "Linear Clamp Sampler",
                                            VkSamplerCreateInfo{
                                                .sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO,
                                                .pNext = nullptr,
                                                .flags = VK_FLAGS_NONE,
                                                .magFilter = VK_FILTER_LINEAR,
                                                .minFilter = VK_FILTER_LINEAR,
                                                .mipmapMode = VK_SAMPLER_MIPMAP_MODE_NEAREST,
                                                .addressModeU = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                .addressModeV = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                .addressModeW = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                .mipLodBias = 0.f,
                                                .anisotropyEnable = VK_FALSE,
                                                .maxAnisotropy = 0,
                                                .compareEnable = VK_FALSE,
                                                .compareOp = VK_COMPARE_OP_ALWAYS,
                                                .minLod = 0.f,
                                                .maxLod = FLT_MAX,
                                                .borderColor = VK_BORDER_COLOR_INT_OPAQUE_BLACK,
                                                .unnormalizedCoordinates = VK_FALSE,
                                            });

    resources.linear_black_border = create_sampler(backend, "Linear Black Border Sampler",
                                                   VkSamplerCreateInfo{
                                                       .sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO,
                                                       .pNext = nullptr,
                                                       .flags = VK_FLAGS_NONE,
                                                       .magFilter = VK_FILTER_LINEAR,
                                                       .minFilter = VK_FILTER_LINEAR,
                                                       .mipmapMode = VK_SAMPLER_MIPMAP_MODE_NEAREST,
                                                       .addressModeU = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                       .addressModeV = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                       .addressModeW = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_EDGE,
                                                       .mipLodBias = 0.f,
                                                       .anisotropyEnable = VK_FALSE,
                                                       .maxAnisotropy = 0,
                                                       .compareEnable = VK_FALSE,
                                                       .compareOp = VK_COMPARE_OP_ALWAYS,
                                                       .minLod = 0.f,
                                                       .maxLod = FLT_MAX,
                                                       .borderColor = VK_BORDER_COLOR_INT_OPAQUE_BLACK,
                                                       .unnormalizedCoordinates = VK_FALSE,
                                                   });

    resources.shadow_map_sampler = create_sampler(
        backend, "Shadow Map Sampler",
        VkSamplerCreateInfo{
            .sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO,
            .pNext = nullptr,
            .flags = VK_FLAGS_NONE,
            .magFilter = VK_FILTER_LINEAR,
            .minFilter = VK_FILTER_LINEAR,
            .mipmapMode = VK_SAMPLER_MIPMAP_MODE_NEAREST,
            .addressModeU = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER,
            .addressModeV = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER,
            .addressModeW = VK_SAMPLER_ADDRESS_MODE_CLAMP_TO_BORDER,
            .mipLodBias = 0.f,
            .anisotropyEnable = VK_FALSE,
            .maxAnisotropy = 0,
            .compareEnable = VK_TRUE,
            .compareOp = ShadowUseReverseZ ? VK_COMPARE_OP_GREATER : VK_COMPARE_OP_LESS,
            .minLod = 0.f,
            .maxLod = FLT_MAX,
            .borderColor = ShadowUseReverseZ ? VK_BORDER_COLOR_INT_OPAQUE_BLACK : VK_BORDER_COLOR_INT_OPAQUE_WHITE,
            .unnormalizedCoordinates = VK_FALSE,
        });

    resources.diffuse_map_sampler = create_sampler(backend, "Diffuse Map Sampler",
                                                   VkSamplerCreateInfo{
                                                       .sType = VK_STRUCTURE_TYPE_SAMPLER_CREATE_INFO,
                                                       .pNext = nullptr,
                                                       .flags = VK_FLAGS_NONE,
                                                       .magFilter = VK_FILTER_LINEAR,
                                                       .minFilter = VK_FILTER_LINEAR,
                                                       .mipmapMode = VK_SAMPLER_MIPMAP_MODE_LINEAR,
                                                       .addressModeU = VK_SAMPLER_ADDRESS_MODE_REPEAT,
                                                       .addressModeV = VK_SAMPLER_ADDRESS_MODE_REPEAT,
                                                       .addressModeW = VK_SAMPLER_ADDRESS_MODE_REPEAT,
                                                       .mipLodBias = 0.f,
                                                       .anisotropyEnable = VK_TRUE,
                                                       .maxAnisotropy = 8,
                                                       .compareEnable = VK_FALSE,
                                                       .compareOp = VK_COMPARE_OP_ALWAYS,
                                                       .minLod = 0.f,
                                                       .maxLod = FLT_MAX,
                                                       .borderColor = VK_BORDER_COLOR_INT_OPAQUE_BLACK,
                                                       .unnormalizedCoordinates = VK_FALSE,
                                                   });

    return resources;
} // namespace Reaper

void destroy_sampler_resources(VulkanBackend& backend, SamplerResources& resources)
{
    vkDestroySampler(backend.device, resources.linear_clamp, nullptr);
    vkDestroySampler(backend.device, resources.linear_black_border, nullptr);
    vkDestroySampler(backend.device, resources.shadow_map_sampler, nullptr);
    vkDestroySampler(backend.device, resources.diffuse_map_sampler, nullptr);
}
} // namespace Reaper

File: src/renderer/vulkan/sampler_resources.zig
// Port of src/renderer/vulkan/SamplerResources.{h,cpp}
//
// NOTE: linear_clamp and linear_black_border are byte-for-byte identical in the
// C++ despite their names — linear_black_border uses CLAMP_TO_EDGE, not
// CLAMP_TO_BORDER. Both are kept so the two names stay available to the passes
// that reference them; fixing the address mode would change what those passes
// sample outside [0,1].

const std = @import("std");
const vk = @import("vulkan");

const shadow_use_reverse_z = @import("../prepare_buckets.zig").shadow_use_reverse_z;

pub const SamplerResources = struct {
    linear_clamp: vk.Sampler,
    linear_black_border: vk.Sampler,
    shadow_map_sampler: vk.Sampler,
    diffuse_map_sampler: vk.Sampler,

    pub fn init(vkd: anytype, device: vk.Device) !SamplerResources {
        const linear_clamp_info = vk.SamplerCreateInfo{
            .s_type = .sampler_create_info,
            .p_next = null,
            .flags = .{},
            .mag_filter = .linear,
            .min_filter = .linear,
            .mipmap_mode = .nearest,
            .address_mode_u = .clamp_to_edge,
            .address_mode_v = .clamp_to_edge,
            .address_mode_w = .clamp_to_edge,
            .mip_lod_bias = 0.0,
            .anisotropy_enable = .false,
            .max_anisotropy = 0,
            .compare_enable = .false,
            .compare_op = .always,
            .min_lod = 0.0,
            .max_lod = std.math.floatMax(f32),
            .border_color = .int_opaque_black,
            .unnormalized_coordinates = .false,
        };

        const linear_clamp = try vkd.createSampler(device, &linear_clamp_info, null);
        errdefer vkd.destroySampler(device, linear_clamp, null);

        // Same create info as above — see the note at the top of the file.
        const linear_black_border = try vkd.createSampler(device, &linear_clamp_info, null);
        errdefer vkd.destroySampler(device, linear_black_border, null);

        const shadow_map_info = vk.SamplerCreateInfo{
            .s_type = .sampler_create_info,
            .p_next = null,
            .flags = .{},
            .mag_filter = .linear,
            .min_filter = .linear,
            .mipmap_mode = .nearest,
            .address_mode_u = .clamp_to_border,
            .address_mode_v = .clamp_to_border,
            .address_mode_w = .clamp_to_border,
            .mip_lod_bias = 0.0,
            .anisotropy_enable = .false,
            .max_anisotropy = 0,
            .compare_enable = .true,
            .compare_op = if (shadow_use_reverse_z) .greater else .less,
            .min_lod = 0.0,
            .max_lod = std.math.floatMax(f32),
            .border_color = if (shadow_use_reverse_z) .int_opaque_black else .int_opaque_white,
            .unnormalized_coordinates = .false,
        };

        const shadow_map_sampler = try vkd.createSampler(device, &shadow_map_info, null);
        errdefer vkd.destroySampler(device, shadow_map_sampler, null);

        const diffuse_map_info = vk.SamplerCreateInfo{
            .s_type = .sampler_create_info,
            .p_next = null,
            .flags = .{},
            .mag_filter = .linear,
            .min_filter = .linear,
            .mipmap_mode = .linear,
            .address_mode_u = .repeat,
            .address_mode_v = .repeat,
            .address_mode_w = .repeat,
            .mip_lod_bias = 0.0,
            .anisotropy_enable = .true,
            .max_anisotropy = 8,
            .compare_enable = .false,
            .compare_op = .always,
            .min_lod = 0.0,
            .max_lod = std.math.floatMax(f32),
            .border_color = .int_opaque_black,
            .unnormalized_coordinates = .false,
        };

        const diffuse_map_sampler = try vkd.createSampler(device, &diffuse_map_info, null);

        return .{
            .linear_clamp = linear_clamp,
            .linear_black_border = linear_black_border,
            .shadow_map_sampler = shadow_map_sampler,
            .diffuse_map_sampler = diffuse_map_sampler,
        };
    }

    pub fn deinit(self: *SamplerResources, vkd: anytype, device: vk.Device) void {
        vkd.destroySampler(device, self.diffuse_map_sampler, null);
        vkd.destroySampler(device, self.shadow_map_sampler, null);
        vkd.destroySampler(device, self.linear_black_border, null);
        vkd.destroySampler(device, self.linear_clamp, null);
    }
};
